Track user connections across clusters

rosstimothy opened the pull request:
Ensure that `proxy_ssh_sessions_total` is incremented in both the root and the leaf cluster if a session to a node in the leaf cluster is established through the root cluster. This now also prevents leaf clusters from terminating during a graceful shutdown if there is an active connection that was created from the root.

Fixes #32605.

fspmarshall commented on lib/reversetunnel/agentpool.go:
```diff
@@ -531,6 +531,14 @@ func (p *AgentPool) transport(ctx context.Context, channel ssh.Channel, requests
 	}
+
+	if p.IsRemoteCluster {
+		if tracker, ok := p.ReverseTunnelServer.(interface{ TrackUserConnection() func() }); ok {
+			t.trackUserConnection = tracker.TrackUserConnection
+		}
```
I find this difficult to parse at a glance... anonymous interface declaring a function that returns a function inside an assignment inside an if statement is just one too many things I think.  I'd suggest declaring this interface separately.

Additionally, I think `TrackUserConnection` is the kind of function that really benefits from named return types:

```golang
type userConnectionTracker interface {
    TrackUserConnection() (cancel func())
}
```

intended usage pattern is much more obvious IMO.

fspmarshall commented on lib/reversetunnel/srv.go:
```diff
@@ -1084,6 +1084,13 @@ func (s *server) rejectRequest(ch ssh.NewChannel, reason ssh.RejectionReason, ms
+// TrackUserConnection tracks a user connection that should prevent
+// the server from being terminated if active. The returned function
+// should be called when the connection is terminated.
+func (s *server) TrackUserConnection() func() {
```
```suggestion
func (s *server) TrackUserConnection() (cancel func()) {
```

fspmarshall commented on lib/reversetunnel/transport.go:
```diff
@@ -93,6 +94,8 @@ type transport struct {
 	// correctly propagate client address in reverse tunnel agents of nodes/services.
 	forwardClientAddress bool
+
+	trackUserConnection func() func()
```
```suggestion
	trackUserConnection func() (cancel func())
```

fspmarshall commented on lib/sshutils/server.go:
```diff
@@ -441,6 +441,17 @@ func (s *Server) trackUserConnections(delta int32) int32 {
+// TrackUserConnection tracks a user connection that should prevent
+// the server from being terminated if active. The returned function
+// should be called when the connection is terminated.
+func (s *Server) TrackUserConnection() func() {
```
```suggestion
func (s *Server) TrackUserConnection() (cancel func()) {
```

espadolini commented on lib/reversetunnel/transport.go:
```diff
@@ -246,6 +250,11 @@ func (p *transport) start() {
 		directAddress = dreq.Address
+
+		if p.trackUserConnection != nil {
+			closed := p.trackUserConnection()
+			defer closed()
```
Why the additional variable?

```suggestion
			defer p.trackUserConnection()()
```

espadolini commented on lib/reversetunnel/agentpool.go:
```diff
@@ -531,6 +531,23 @@ func (p *AgentPool) transport(ctx context.Context, channel ssh.Channel, requests
+	// the user tracker the leaf cluster metrics will be incorrect and graceful shutdown will not
+	// wait for user sessions to be terminated prior to proceeding with the shutdown operation.
+	if p.IsRemoteCluster {
+		type userConnectionTracker interface {
+			TrackUserConnection() (closed func())
+		}
```
I tried adding `TrackUserConnection` to `reversetunnelclient.Server` and it seems to work. Why the additional interface?

espadolini commented on lib/reversetunnel/transport.go:
```diff
@@ -93,6 +94,9 @@ type transport struct {
 	forwardClientAddress bool
+
+	// trackUserConnection is an optional mechanism used to count active user sessions.
+	trackUserConnection func() (closed func())
```
IMO

```suggestion
	trackUserConnection func() (release func())
```

rosstimothy committed:
Track user connections across clusters

Ensure that `proxy_ssh_sessions_total` is incremented in both the
root and the leaf cluster if a session to a node in the leaf cluster
is established through the root cluster. This now also prevents
leaf clusters from terminating during a graceful shutdown if there
is an active connection that was created from the root.

Fixes #32605.
